### PR TITLE
feat(pubsub): add bulk subscribe support

### PR DESCRIPTION
**Pull request by shubham1172**
# Description

**NOTE**: This PR should be only merged once the bulk pubsub feature (2218) is checked-in Dapr. The Dapr reference in build/validate YAMLs must be updated back to `master`.

Add a new annotation `@BulkSubscribe` to allow subscribing to events using the Bulk Subscribe API.

## Issue reference

Please reference the issue this PR will close: #778 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

### CODE REVIEW

File: sdk-springboot/src/main/java/io/dapr/springboot/domain/DaprBulkAppResponse.java
Comment (mukundansundar) [inline sdk-springboot/src/main/java/io/dapr/springboot/domain/DaprBulkAppResponse.java:21]: all fields should be private with get accessors only ...

Comment (mukundansundar) [reply sdk-springboot/src/main/java/io/dapr/springboot/domain/DaprBulkAppResponse.java:21]: This domain object should be part of default SDK and not springboot ...

**Review by mukundansundar — DISMISSED** [review]

File: sdk-springboot/src/main/java/io/dapr/springboot/domain/DaprBulkAppResponse.java
Comment (shubham1172) [reply sdk-springboot/src/main/java/io/dapr/springboot/domain/DaprBulkAppResponse.java:21]: Thanks, updated!

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java
Comment (pravinpushkar) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java:40]: Is this necessary ? If not can we keep it private only ?

File: sdk/src/main/java/io/dapr/client/domain/DaprBulkAppResponse.java
Comment (pravinpushkar) [inline sdk/src/main/java/io/dapr/client/domain/DaprBulkAppResponse.java:35]: Is it possible to use List here ?

Comment (shubham1172) [reply sdk/src/main/java/io/dapr/client/domain/DaprBulkAppResponse.java:35]: Kept it as an array to not force users to use `List` and keep the contract primitive. Users are still free to use a `List` and use `myList.toArray()` to create a `DaprBulkAppResponse`.

**Comment by pravinpushkar** [comment]
@shubham1172 The build failure is due to a diff reason. I have the fix in this PR - https://github.com/dapr/java-sdk/pull/740

File: examples/src/main/java/io/dapr/examples/pubsub/http/README.md
Comment (pravinpushkar) [inline examples/src/main/java/io/dapr/examples/pubsub/http/README.md]: unexpected qaQ

Comment (shubham1172) [reply examples/src/main/java/io/dapr/examples/pubsub/http/README.md]: Removed

**Comment by shubham1172** [comment]
Do not merge until dapr/mechanical-markdown#29 is merged and tests have been added here.

**Comment by shubham1172** [comment]
@mukundansundar @pravinpushkar this is ready for review.

**Comment by mukundansundar** [comment]
@shubham1172 protos and code should be updated after https://github.com/dapr/dapr/pull/5489 is merged

**Comment by pravinpushkar** [comment]
LGTM..however, The PR checks for method Invoke failing.

**Comment by shubham1172** [comment]
> LGTM..however, The PR checks for method Invoke failing.

Yes, I need to troubleshoot that - over the top of your head can you share any potential reasons for this failure?

File: sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java
Comment (shubham1172) [inline sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java:144]: Unrelated to this PR, but see @mukundansundar's comment from #789 

> related to https://github.com/dapr/dapr/pull/5648/files#diff-ce3f68305f0eb28b6b8911c281186536c80d8cac9dff3cece2c8bdf3dd955315R465-R475
now we get in gRPC (io.dapr.exceptions.DaprException: INTERNAL: fail to invoke, id: methodinvokeit-methodinvokeservice, err: message is nil)
instead of UNKNOWN

**Comment by mukundansundar** [comment]
@shubham1172 One point, potentially `BulkPublishEntry` in #789 and `DaprBulkMessageEntry` can be combined. Based on which PR goes in first both domain objects can be combined. 
Also do we need to prefix with Dapr on the domain objects? Like `DaprBulkMessage` ? Can it just be `BulkMessage`?

**Comment by shubham1172** [comment]
> @shubham1172 One point, potentially `BulkPublishEntry` in #789 and `DaprBulkMessageEntry` can be combined. Based on which PR goes in first both domain objects can be combined. Also do we need to prefix with Dapr on the domain objects? Like `DaprBulkMessage` ? Can it just be `BulkMessage`?

Thanks Mukundan, I have removed the `Dapr` prefix from domain objects.

File: examples/src/main/java/io/dapr/examples/pubsub/http/README.md
Comment (artursouza) [inline examples/src/main/java/io/dapr/examples/pubsub/http/README.md:63]: Show the example of BulkSubscribe after the regular subscribe.

File: pom.xml
Comment (artursouza) [inline pom.xml:19]: Not master, please. Just pin the commit ref.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java
Comment (artursouza) [reply sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java:39]: Please, use reflection in the unit test to instantiate this class and keep it private. This is important, otherwise users can mess this up.

This blog post describes how to do it: https://avaldes.com/how-do-i-instantiate-a-class-with-a-private-constructor/

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java:68]: Create an overloaded method with the previous signature too to avoid a breaking change.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprSubscriptionBuilder.java:95]: Return the `this` here too to keep the fluent pattern.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java:28]: final

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java
Comment (artursouza) [inline sdk-springboot/src/main/java/io/dapr/springboot/DaprTopicSubscription.java:49]: Add an overloaded method to keep the previous signature and backwards compatibility.

File: sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java
Comment (artursouza) [inline sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java:39]: You can undo this after adding the overloaded method.

File: sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java
Comment (artursouza) [inline sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java:41]: You can undo this after adding the overloaded method.

File: sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java
Comment (artursouza) [inline sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java:73]: You can undo this after adding the overloaded method.

File: sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java
Comment (artursouza) [inline sdk-springboot/src/test/java/io/dapr/springboot/DaprRuntimeTest.java:77]: You can undo this after adding the overloaded method.

File: sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java
Comment (artursouza) [inline sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java:140]: I am looking into this too.

File: sdk/src/main/java/io/dapr/client/domain/BulkAppResponse.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkAppResponse.java:23]: Sorry for this but make sure this name is specific to PubSub. The current name is too generic.

File: sdk/src/main/java/io/dapr/client/domain/BulkAppResponseEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkAppResponseEntry.java:19]: Same regarding name.

File: sdk/src/main/java/io/dapr/client/domain/BulkAppResponseEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkAppResponseEntry.java:43]: Can this class be immutable?

File: sdk/src/main/java/io/dapr/client/domain/BulkAppResponse.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkAppResponse.java:44]: Can this class be immutable?

File: sdk/src/main/java/io/dapr/client/domain/BulkMessage.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkMessage.java:31]: Can you really instantiate  BulkMessage with nothing?

File: sdk/src/main/java/io/dapr/client/domain/BulkMessage.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkMessage.java:50]: Can this class be immutable?

File: sdk/src/main/java/io/dapr/client/domain/BulkMessageEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkMessageEntry.java:32]: same.

File: sdk/src/main/java/io/dapr/client/domain/BulkMessageEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkMessageEntry.java:61]: Immutable?

File: sdk/src/test/java/io/dapr/client/domain/BulkAppResponseEntryTest.java
Comment (artursouza) [inline sdk/src/test/java/io/dapr/client/domain/BulkAppResponseEntryTest.java:22]: I am not a fan of testing domain classes, ideally the test of the actual functions should cover those. Please, delete these unit tests.

File: sdk/src/test/java/io/dapr/client/domain/BulkAppResponseTest.java
Comment (artursouza) [inline sdk/src/test/java/io/dapr/client/domain/BulkAppResponseTest.java:25]: Same.

File: sdk/src/test/java/io/dapr/client/domain/BulkMessageEntryTest.java
Comment (artursouza) [inline sdk/src/test/java/io/dapr/client/domain/BulkMessageEntryTest.java:23]: Same.

File: sdk/src/test/java/io/dapr/client/domain/BulkMessageTest.java
Comment (artursouza) [inline sdk/src/test/java/io/dapr/client/domain/BulkMessageTest.java:27]: Same.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml
Comment (shubham1172) [reply .github/workflows/build.yml:32]: Will we not only merge this PR once a RC is cut in dapr/dapr? If we can merge it now, I can replace this with a commit.

File: examples/src/main/java/io/dapr/examples/pubsub/http/Publisher.java
Comment (shubham1172) [reply examples/src/main/java/io/dapr/examples/pubsub/http/Publisher.java:43]: We would still need different topics since there is only once instance of subscriber running with a single app-id. Have refactored publisher to use a topic from `args` and fallback to a default otherwise.

File: sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java
Comment (shubham1172) [reply sdk-springboot/src/main/java/io/dapr/springboot/DaprRuntime.java:39]: Since the scope is `default`, wouldn't it disallow users from accessing it? For now I have reverted it back to private and used reflection in test.

File: sdk/src/main/java/io/dapr/client/domain/BulkMessage.java
Comment (shubham1172) [reply sdk/src/main/java/io/dapr/client/domain/BulkMessage.java:31]: I am taking this implementation from #789 and making it immutable.

File: sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java
Comment (artursouza) [reply sdk-tests/src/test/java/io/dapr/it/methodinvoke/grpc/MethodInvokeIT.java:140]: Resolve the conflicts from master, please. For that, you can just pick the master changes.

File: sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessage.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessage.java:37]: Sorry for this but we should not have dependency on Jackson in our domain objects. Now, it seems that this is only required at the sdk-springboot. So, I would create an inherited class from his one with the same name but in a java package called `io.dapr.client.domain.jackson` in sdk-springboot project. In the docs, you make it clear that the jackson package should be used instead.

Another options (which might be better) is to make all these classes specific to BulkSubscribe and remove any "PubSub" from the name in favor of "Subscribe", moving all of it to the sdk-springboot project in the `io.dapr.client.domain.jackson` package, avoiding any duplication of code.

File: sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessageEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessageEntry.java:43]: Same. No dependency on Jackson in our APIs.

File: sdk/src/main/java/io/dapr/client/domain/BulkSubscribeAppResponse.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkSubscribeAppResponse.java:32]: Same.

File: sdk/src/main/java/io/dapr/client/domain/BulkSubscribeAppResponseEntry.java
Comment (artursouza) [inline sdk/src/main/java/io/dapr/client/domain/BulkSubscribeAppResponseEntry.java:34]: Same.

**Review by artursouza — CHANGES_REQUESTED** [review]

**Comment by artursouza** [comment]
Also, resolve conflicts from master, please.

File: sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessage.java
Comment (shubham1172) [reply sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessage.java:37]: FYI we do depend on Jackson already in domain objects, example -

https://github.com/dapr/java-sdk/blob/b83661d1094361a6c41a9681d903a4d064687106/sdk/src/main/java/io/dapr/client/domain/SubscribeConfigurationResponse.java#L43-L49

Comment (shubham1172) [reply sdk/src/main/java/io/dapr/client/domain/BulkPubSubMessage.java]: As discussed offline, skipping this change for now. Only renamed `PubSub` => `Subscribe`.

**Review by artursouza — APPROVED** [review]

**Pull request merged**
